internal/cmd/envoy/shutdown_manager.go

```diff
@@ -0,0 +1,234 @@
+		if err := srv.Shutdown(context.Background()); err != nil {
+			logger.Error(err, "server shutdown error")
+		}
+		close(c)
+	}()
+
+	// Start HTTP server
+	logger.Info("starting shutdown manager")
+	if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+		logger.Error(err, "starting shutdown manager failed")
+	}
+
+	// Wait until done
+	<-c
+	return nil
+}
+
+// shutdownReadyHandler handles the endpoint used by a preStop hook on the Envoy
+// container to block until ready to terminate. When the graceful drain process
+// has begun a file will be written to indicate shutdown is in progress. This
+// handler will poll for the file and block until it is removed.
+func shutdownReadyHandler(w http.ResponseWriter, readyTimeout time.Duration, activeFile string) {
+	var startTime = time.Now()
+
+	logger.Info("received shutdown ready request")
+
+	// Since we cannot establish order between preStop hooks, wait a bit giving the exec
+	// preStop hook a chance to touch the signaling file before starting to poll for it.
+	time.Sleep(2 * time.Second)
+
+	// Exit early if active file failed to show up
+	if _, err := os.Stat(activeFile); os.IsNotExist(err) {
+		logger.Info("shutdown active file not found")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Poll for shutdown readiness
+	for {
+		// Check if ready timeout is exceeded
+		elapsedTime := time.Since(startTime)
+		if elapsedTime > readyTimeout {
+			logger.Info("shutdown readiness timeout exceeded")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		_, err := os.Stat(activeFile)
+		switch {
+		case os.IsNotExist(err):
+			logger.Info("shutdown readiness detected")
+			return
+		case err != nil:
+			logger.Error(err, "error checking for shutdown readiness")
+		case err == nil:
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+// Shutdown is called from a preStop hook on the shutdown-manager container where
+// it will initiate a graceful drain sequence on the Envoy proxy and block until
+// connections are drained or a timeout is exceeded.
+func Shutdown(drainTimeout time.Duration, minDrainDuration time.Duration, exitAtConnections int) error {
+	var startTime = time.Now()
+	var allowedToExit = false
+
+	// Reconfigure logger to write to stdout of main process if running in Kubernetes
+	if _, k8s := os.LookupEnv("KUBERNETES_SERVICE_HOST"); k8s && os.Getpid() != 1 {
+		logger = logging.FileLogger("/proc/1/fd/1", "shutdown-manager", v1alpha1.LogLevelInfo)
+	}
+
+	// Signal to shutdownReadyHandler that drain process is started
+	if _, err := os.Create(ShutdownActiveFile); err != nil {
+		logger.Error(err, "error creating shutdown active file")
+		return err
+	}
+
+	logger.Info(fmt.Sprintf("initiating graceful drain with %.0f second minimum drain period and %.0f second timeout",
+		minDrainDuration.Seconds(), drainTimeout.Seconds()))
+
+	// Send request to Envoy admin API to initiate the graceful drain sequence
```

Comment:
would it make more sense for shutdown logic to be moved into `shutdownReadyHandler`

so the shutdown command can
* create a file
* wait until file is deleted

Reply:
I strongly believe the logic here should be kept in an `exec` preStop. There is no reason really to push this into the HTTP server, and doing so opens the door to multiple drain processes being triggered once the file is in place. Unless there is a significant reason behind the nudge here, I'd like to keep it as-is.

Reply:
I see, you'd prefer if the `shutdown` cmd performs the shutdown logic, while the http handler performs the simpler block in case the file exists, that sgtm